Amazon Complaining about Missing Date Header When Creating/Fetching S3 Repository

I'm currently using v2.5.1 of this plugin successfully on one of my servers (running inside the official ES Docker image) to create and snapshot to an S3 repository. I recently started a new instance of ES in a new Docker container on the same server with the same exact configuration with the intention of restoring a snapshot, but noticed I could not create repositories or fetch listings of snapshots that exist in those repositories from the new instance. On the original instance, everything still works as expected, but now I'm scared to shut down that Docker container since it is working correctly at the moment. As a side note, I did this exact same process just a couple days ago with success, so I'm not sure what has changed since Wednesday.

This is the error I'm getting:

```json
{
   "error": "RemoteTransportException[[server][inet[/192.168.1.1:9300]][cluster:admin/repository/put]]; nested: RepositoryVerificationException[[my_s3_repository] path [my_s3_path] is not accessible on master node]; nested: IOException[Unable to upload object my_s3_path/tests-ooBzS8deRtWxFeS6OuG9SQ-master]; nested: AmazonS3Exception[AWS authentication requires a valid Date or x-amz-date header (Service: Amazon S3; Status Code: 403; Error Code: AccessDenied; Request ID: C810733A376C4D79)]; ",
   "status": 500
}
```

Any idea what would be causing this? Thanks in advance!

It seems the required change is to use Amazon SDK v1.10.11 for any version of Java 1.8u60 and above (in the case of the base Docker images, 1.8u66 is the new base version used for 1.8).

I'm not exactly sure how to go about testing this or if there is something else that needs to be done to get this change into all of the versions, but I tried modifying the pom.xml file in v1.5 branch and run a "mvn clean install", but get the following error:

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.2:testCompile (default-testCompile) on project elasticsearch-cloud-aws: Compilation failure
[ERROR] /Users/marcmelvin/Development/SmartProcure/elasticsearch-cloud-aws/src/test/java/org/elasticsearch/cloud/aws/AmazonS3Wrapper.java:[41,7] error: AmazonS3Wrapper is not abstract and does not override abstract method getBucketReplicationConfiguration(GetBucketReplicationConfigurationRequest) in AmazonS3

@lusid Thats because that class no longer implements all the methods present in the AmazonS3.

I've actually implemented them, packaged a new version with the latest Amazon SDK dependency but it still gives the same error.

Upon inspecting the generated zip, I saw that it still includes `joda-time 1.7.0`. This comes from `elasticsearch-parent` project [[pom]](http://central.maven.org/maven2/org/elasticsearch/elasticsearch-parent/1.7.0/elasticsearch-parent-1.7.0.pom).

Forcing to use the latest joda-time will issue an runtime error on ES:
```
1) Error injecting constructor, java.lang.NoClassDefFoundError: org/joda/time/ReadableInstant
  at org.elasticsearch.repositories.s3.S3Repository.<init>(Unknown Source)
  while locating org.elasticsearch.repositories.s3.S3Repository
  while locating org.elasticsearch.repositories.Repository

1 error
	at org.elasticsearch.common.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:344)
	at org.elasticsearch.common.inject.InjectorBuilder.injectDynamically(InjectorBuilder.java:178)
	at org.elasticsearch.common.inject.InjectorBuilder.build(InjectorBuilder.java:110)
	at org.elasticsearch.common.inject.InjectorImpl.createChildInjector(InjectorImpl.java:131)
	at org.elasticsearch.common.inject.ModulesBuilder.createChildInjector(ModulesBuilder.java:69)
	at org.elasticsearch.repositories.RepositoriesService.createRepositoryHolder(RepositoriesService.java:407)
	at org.elasticsearch.repositories.RepositoriesService.clusterChanged(RepositoriesService.java:302)
	at org.elasticsearch.cluster.service.InternalClusterService$UpdateTask.run(InternalClusterService.java:480)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.runAndClean(PrioritizedEsThreadPoolExecutor.java:196)
	at org.elasticsearch.common.util.concurrent.PrioritizedEsThreadPoolExecutor$TieBreakingPrioritizedRunnable.run(PrioritizedEsThreadPoolExecutor.java:162)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.NoClassDefFoundError: org/joda/time/ReadableInstant
	at com.amazonaws.services.s3.internal.ServiceUtils.<clinit>(ServiceUtils.java:67)
	at com.amazonaws.services.s3.internal.S3Signer.sign(S3Signer.java:168)
(...)
```

I think Elasticsearch is not loading the joda-time present in the plugin, so I think Elasticsearch's joda-time needs to be updated to use the latest version.

--- a/pom.xml
+++ b/pom.xml
@@ -32,7 +32,7 @@
     </parent>
 
     <properties>
-        <amazonaws.version>1.9.34</amazonaws.version>
+        <amazonaws.version>1.10.11</amazonaws.version>
         <tests.jvms>1</tests.jvms>
     </properties>
 

--- a/src/test/java/org/elasticsearch/cloud/aws/AmazonS3Wrapper.java
+++ b/src/test/java/org/elasticsearch/cloud/aws/AmazonS3Wrapper.java
@@ -579,4 +579,78 @@ public class AmazonS3Wrapper implements AmazonS3 {
     public boolean isRequesterPaysEnabled(String bucketName) throws AmazonServiceException, AmazonClientException {
         return delegate.isRequesterPaysEnabled(bucketName);
     }
+
+
+	@Override
+	public ObjectListing listNextBatchOfObjects(ListNextBatchOfObjectsRequest listNextBatchOfObjectsRequest)
+			throws AmazonClientException, AmazonServiceException {
+		return delegate.listNextBatchOfObjects(listNextBatchOfObjectsRequest);
+	}
+
+
+	@Override
+	public VersionListing listNextBatchOfVersions(ListNextBatchOfVersionsRequest listNextBatchOfVersionsRequest)
+			throws AmazonClientException, AmazonServiceException {
+		return delegate.listNextBatchOfVersions(listNextBatchOfVersionsRequest);
+	}
+
+
+	@Override
+	public Owner getS3AccountOwner(GetS3AccountOwnerRequest getS3AccountOwnerRequest)
+			throws AmazonClientException, AmazonServiceException {
+		return delegate.getS3AccountOwner(getS3AccountOwnerRequest);
+	}
+
+
+	@Override
+	public BucketLoggingConfiguration getBucketLoggingConfiguration(
+			GetBucketLoggingConfigurationRequest getBucketLoggingConfigurationRequest)
+					throws AmazonClientException, AmazonServiceException {
+		return delegate.getBucketLoggingConfiguration(getBucketLoggingConfigurationRequest);
+	}
+
+
+	@Override
+	public BucketVersioningConfiguration getBucketVersioningConfiguration(
+			GetBucketVersioningConfigurationRequest getBucketVersioningConfigurationRequest)
+					throws AmazonClientException, AmazonServiceException {
+		return delegate.getBucketVersioningConfiguration(getBucketVersioningConfigurationRequest);
+	}
+
+
+	@Override
+	public BucketLifecycleConfiguration getBucketLifecycleConfiguration(
+			GetBucketLifecycleConfigurationRequest getBucketLifecycleConfigurationRequest) {
+		return delegate.getBucketLifecycleConfiguration(getBucketLifecycleConfigurationRequest);
+	}
+
+
+	@Override
+	public BucketCrossOriginConfiguration getBucketCrossOriginConfiguration(
+			GetBucketCrossOriginConfigurationRequest getBucketCrossOriginConfigurationRequest) {
+		return delegate.getBucketCrossOriginConfiguration(getBucketCrossOriginConfigurationRequest);
+	}
+
+
+	@Override
+	public BucketTaggingConfiguration getBucketTaggingConfiguration(
+			GetBucketTaggingConfigurationRequest getBucketTaggingConfigurationRequest) {
+		return delegate.getBucketTaggingConfiguration(getBucketTaggingConfigurationRequest);
+	}
+
+
+	@Override
+	public BucketNotificationConfiguration getBucketNotificationConfiguration(
+			GetBucketNotificationConfigurationRequest getBucketNotificationConfigurationRequest)
+					throws AmazonClientException, AmazonServiceException {
+		return delegate.getBucketNotificationConfiguration(getBucketNotificationConfigurationRequest);
+	}
+
+
+	@Override
+	public BucketReplicationConfiguration getBucketReplicationConfiguration(
+			GetBucketReplicationConfigurationRequest getBucketReplicationConfigurationRequest)
+					throws AmazonServiceException, AmazonClientException {
+		return delegate.getBucketReplicationConfiguration(getBucketReplicationConfigurationRequest);
+	}
 }

